core/src/main/java/jenkins/security/BasicHeaderApiTokenAuthenticator.java

```diff
@@ -21,14 +22,23 @@
  */
 @Extension
 public class BasicHeaderApiTokenAuthenticator extends BasicHeaderAuthenticator {
+    /**
+     * Note: if the token does not exist or does not match, we do not use {@link SecurityListener#fireFailedToAuthenticate(String)}
+     * because it will be done in the {@link BasicHeaderRealPasswordAuthenticator} in the case the password is not valid either
+     */
     @Override
     public Authentication authenticate(HttpServletRequest req, HttpServletResponse rsp, String username, String password) throws ServletException {
         // attempt to authenticate as API token
         User u = User.getById(username, true);
         ApiTokenProperty t = u.getProperty(ApiTokenProperty.class);
         if (t!=null && t.matchesPassword(password)) {
             try {
-                return u.impersonate();
+                UserDetails userDetails = u.getUserDetailsForImpersonation();
+                Authentication auth = u.impersonate(userDetails);
+
+                SecurityListener.fireAuthenticated(userDetails);
```

Comment:
Makes [this point of Javadoc](https://github.com/jenkinsci/jenkins/blob/bcbd5ba56ad7bbed00977b2d733f60019c6af04c/core/src/main/java/jenkins/security/SecurityListener.java#L49) inaccurate.